### PR TITLE
Enabling typedef tslint rule for taco packages

### PR DESCRIPTION
**Pull request by nisheetjain**

### CODE REVIEW

File: src/remotebuild/lib/darwin/darwinCerts.ts [hunk 1]
```diff
@@ -28,11 +28,10 @@ import tacoUtils = require ("taco-utils");
 
 import utils = tacoUtils.UtilHelper;
 import logger = tacoUtils.Logger;
-var exec = child_process.exec;
 
 class Certs {
-    private static debug = false;
-    private static CERT_DEFAULTS = {
+    private static debug: boolean = false;
+    private static CERT_DEFAULTS: any = {
```
Comment (digeff) [inline src/remotebuild/lib/darwin/darwinCerts.ts]: Please put the proper type for this variable. With any, you can use any members, like dayss and the compiler will allow it. If you leave it untyped, typescript deduces the proper type, so trying to use Certs.CERT_DEFAULTS.dayss gives this error:
remotebuild\lib\darwin\darwinCerts.ts(227,64): error TS2339: Property 'dayss' does not exist on type '{ days: number; country: string; ca_cn: string; pfx_name: string; client_cn: string; }'.

File: src/remotebuild/lib/darwin/darwinCerts.ts [hunk 2]
```diff
@@ -96,8 +95,8 @@ class Certs {
     }
 
     public static initializeServerCerts(conf: RemoteBuildConf): Q.Promise<HostSpecifics.ICertStore> {
-        var certsDir = path.join(conf.serverDir, "certs");
-        var certPaths = {
+        var certsDir: string = path.join(conf.serverDir, "certs");
+        var certPaths: any = {
```
Comment (digeff) [inline src/remotebuild/lib/darwin/darwinCerts.ts]: See previous comment. Please use a proper type here.

File: src/remotebuild/lib/server.ts [hunk 3]
```diff
@@ -238,7 +238,7 @@ class Server {
                 return Q(certStore);
             }).
             then(function (certStore: HostSpecifics.ICertStore): https.Server {
-                var sslSettings = {
+                var sslSettings: any = {
```
Comment (digeff) [inline src/remotebuild/lib/server.ts]: Same comment as before. Please use proper type.

File: src/remotebuild/test/darwinCerts.ts [hunk 4]
```diff
@@ -17,7 +17,7 @@
 /* tslint:disable:no-var-requires */
 // var require needed for should module to work correctly
 // Note not import: We don't want to refer to shouldModule, but we need the require to occur since it modifies the prototype of Object.
-var shouldModule = require("should");
+var shouldModule: any = require("should");
```
Comment (digeff) [inline src/remotebuild/test/darwinCerts.ts]: If we are not going to use it, shouldn't we just do:
require("should"); ?

File: src/remotebuild/test/darwinCerts.ts [hunk 5]
```diff
@@ -116,7 +116,7 @@ macOnly("Certs", function (): void {
 
     // Tests that resetting server certificates queries the user and respects a "no" response
     it("ResetServerCertNo", function (done: MochaDone): void {
-        var noHandler = {
+        var noHandler: any = {
```
Comment (digeff) [inline src/remotebuild/test/darwinCerts.ts]: Same comment as before, please use proper type.

File: src/remotebuild/test/darwinCerts.ts [hunk 6]
```diff
@@ -145,7 +145,7 @@ macOnly("Certs", function (): void {
 
     // Tests that resetting server certificates will create new certificates after a "yes" response
     it("ResetServerCertYes", function (done: MochaDone): void {
-        var yesHandler = {
+        var yesHandler: any = {
```
Comment (digeff) [inline src/remotebuild/test/darwinCerts.ts]: Same comment as before, please use proper type.

File: src/remotebuild/test/server.ts [hunk 7]
```diff
@@ -14,7 +14,7 @@
 /* tslint:disable:no-var-requires */
 // var require needed for should module to work correctly
 // Note not import: We don't want to refer to shouldModule, but we need the require to occur since it modifies the prototype of Object.
-var shouldModule = require("should");
+var shouldModule: any = require("should");
```
Comment (digeff) [inline src/remotebuild/test/server.ts]: Same comment as before, should we just use the require here?

File: src/remotebuild/test/darwinCerts.ts (hunk 4 above)
Comment (MSLaguana) [inline src/remotebuild/test/darwinCerts.ts]: At one point that resulted in typescript removing the require statement completely because it was "obviously" unused, which is why we went to this approach. They may have fixed their behaviour since then, I'm not sure.

File: src/taco-cli/cli/build.ts [hunk 8]
```diff
@@ -47,7 +46,7 @@ class Build extends commands.TacoCommandBase {
     /*
      * Exposed for testing purposes: when we talk to a mocked server we don't want 5s delays between pings
      */
-    public static remoteBuild = RemoteBuildClientHelper;
+    public static remoteBuild: any = RemoteBuildClientHelper;
```
Comment (digeff) [inline src/taco-cli/cli/build.ts]: The type of this variable should be: "typeof RemoteBuildClientHelper"

File: src/taco-cli/cli/kit.ts [hunk 9]
```diff
@@ -436,7 +436,7 @@ class Kit extends commands.TacoCommandBase {
         assert(componentType === ProjectComponentType.Platform || componentType === ProjectComponentType.Plugin);
         var componentUpdates: IDictionary<string> = {};
         return kitHelper.getKitInfo(kitId).then(function (kitInfo: TacoKits.IKitInfo): Q.Promise<any> {
-            var componentOverrides = (componentType === ProjectComponentType.Platform) ? kitInfo.platforms : kitInfo.plugins;
+            var componentOverrides: any = (componentType === ProjectComponentType.Platform) ? kitInfo.platforms : kitInfo.plugins;
```
Comment (digeff) [inline src/taco-cli/cli/kit.ts]: Proper type of this variable is TacoKits.IPluginOverrideMetadata | TacoKits.IPlatformOverrideMetadata

File: src/taco-cli/cli/remoteBuild/remoteBuildClientHelper.ts [hunk 10]
```diff
@@ -148,8 +148,8 @@ class RemoteBuildClientHelper {
     }
 
     public static emulate(buildInfo: BuildInfo, serverSettings: Settings.IRemoteConnectionInfo, target: string): Q.Promise<BuildInfo> {
-        var buildUrlBase = Settings.getRemoteServerUrl(serverSettings) + "/build/" + buildInfo.buildNumber;
-        var httpSettings = { language: buildInfo.buildLang, agent: ConnectionSecurityHelper.getAgent(serverSettings) };
+        var buildUrlBase: string = Settings.getRemoteServerUrl(serverSettings) + "/build/" + buildInfo.buildNumber;
+        var httpSettings: { language: string; agent: Q.Promise<https.Agent> } = { language: buildInfo.buildLang, agent: ConnectionSecurityHelper.getAgent(serverSettings) };
```
Comment (digeff) [inline src/taco-cli/cli/remoteBuild/remoteBuildClientHelper.ts]: You are using this type 3 times. Consider declaring:
type HttpSettings = { language: string; agent: Q.Promise<https.Agent> };
and using
httpSettings: HttpSettings

File: src/taco-cli/cli/utils/cordovaHelper.ts [hunk 11]
```diff
@@ -25,7 +25,7 @@ import packageLoader = tacoUtility.TacoPackageLoader;
 class CordovaHelper {
     private static CORDOVA_NPM_PACKAGE_NAME: string = "cordova";
     // Cordova's known parameters
-    private static CORDOVA_BOOLEAN_PARAMETERS =
+    private static CORDOVA_BOOLEAN_PARAMETERS: any =
```
Comment (digeff) [inline src/taco-cli/cli/utils/cordovaHelper.ts]: Same comment, please use complete type.

File: src/taco-cli/cli/utils/cordovaHelper.ts [hunk 12]
```diff
@@ -44,7 +44,7 @@ class CordovaHelper {
         nobuild: Boolean,
         list: Boolean
     };
-    private static CORDOVA_VALUE_PARAMETERS =
+    private static CORDOVA_VALUE_PARAMETERS: any =
```
Comment (digeff) [inline src/taco-cli/cli/utils/cordovaHelper.ts]: Same comment, please use complete type.

File: src/taco-cli/cli/utils/projectHelper.ts [hunk 13]
```diff
@@ -244,9 +244,9 @@ class ProjectHelper {
         .then(function (pluginsInstalled: string[]): Q.Promise<any> {
             pluginsInstalled.forEach(function (plugin: string): void {
                 // Ignore plugins without a package.json
-                var pluginPackgeJson = path.join(projectDir, "plugins", plugin, "package.json");
+                var pluginPackgeJson: string = path.join(projectDir, "plugins", plugin, "package.json");
                 if (fs.existsSync(pluginPackgeJson)) {
-                    var pluginInfo = require(pluginPackgeJson);
+                    var pluginInfo: any = require(pluginPackgeJson);
```
Comment (digeff) [inline src/taco-cli/cli/utils/projectHelper.ts]: This type should be cordovaHelper.IDictionary<string>

File: src/taco-cli/cli/utils/projectHelper.ts [hunk 14]
```diff
@@ -292,7 +292,7 @@ class ProjectHelper {
         // JsonSerializer class in the taco-utils does the necessary formatting
         // This is important as VS expects the JSON file to be in the standard JSON format
         var jsonSerializer: tacoUtility.JsonSerializer = new tacoUtility.JsonSerializer();
-        var formattedTacoJson = jsonSerializer.serialize(jsonData);
+        var formattedTacoJson: any = jsonSerializer.serialize(jsonData);
```
Comment (digeff) [inline src/taco-cli/cli/utils/projectHelper.ts]: This should be a string

File: src/taco-cli/test/build.ts [hunk 15]
```diff
@@ -15,7 +15,7 @@
 /* tslint:disable:no-var-requires */
 // var require needed for should module to work correctly
 // Note not import: We don't want to refer to shouldModule, but we need the require to occur since it modifies the prototype of Object.
-var shouldModule = require("should");
+var shouldModule: any = require("should");
```
Comment (digeff) [inline src/taco-cli/test/build.ts]: Same comment as before. Can we call require directly?

File: src/taco-cli/test/build.ts [hunk 16]
```diff
@@ -72,7 +72,7 @@ describe("taco build", function (): void {
         });
     }
 
-    var remoteServerConfiguration = { host: "localhost", port: 3000, secure: false, mountPoint: "cordova" };
+    var remoteServerConfiguration: any = { host: "localhost", port: 3000, secure: false, mountPoint: "cordova" };
```
Comment (digeff) [inline src/taco-cli/test/build.ts]: Same comment as before, please user complete type.

File: src/taco-cli/test/build.ts [hunk 17]
```diff
@@ -124,12 +124,12 @@ describe("taco build", function (): void {
...
-        var buildArguments = ["--remote", "test"];
-        var configuration = "debug";
-        var buildNumber = 12340;
+        var buildArguments: string [] = ["--remote", "test"];
+        var configuration: string = "debug";
+        var buildNumber: number = 12340;
 
         // Mock out the server on the other side
-        var sequence = [
+        var sequence: any = [
```
Comment (digeff) [inline src/taco-cli/test/build.ts]: Same comment as before, please user complete type.

File: src/remotebuild/test/darwinCerts.ts (hunk 4 above)
Comment (digeff) [inline src/remotebuild/test/darwinCerts.ts]: I didn't know that. According to this issue: https://github.com/Microsoft/TypeScript/issues/596 it seems that the correct way to handle this might be:
import "should";  // just for side effect

File: src/taco-cli/test/build.ts [hunk 18]
```diff
@@ -377,7 +377,7 @@ describe("taco build", function (): void {
             }
         ];
 
-        var serverFunction = ServerMock.generateServerFunction(mocha, sequence);
+        var serverFunction: (request: http.ServerRequest, response: http.ServerResponse) => void = ServerMock.generateServerFunction(mocha, sequence);
```
Comment (digeff) [inline src/taco-cli/test/build.ts]: This type seems quite common. Maybe we should create a specific alias or type

File: src/taco-cli/cli/kit.ts (hunk 9 above)
Comment (nisheetjain) [inline src/taco-cli/cli/kit.ts]: done

File: src/taco-cli/cli/remoteBuild/remoteBuildClientHelper.ts (hunk 10 above)
Comment (nisheetjain) [inline src/taco-cli/cli/remoteBuild/remoteBuildClientHelper.ts]: done

File: src/taco-cli/cli/utils/projectHelper.ts (hunk 13 above)
Comment (nisheetjain) [inline src/taco-cli/cli/utils/projectHelper.ts]: done

File: src/taco-cli/cli/utils/projectHelper.ts (hunk 14 above)
Comment (nisheetjain) [inline src/taco-cli/cli/utils/projectHelper.ts]: done

File: src/taco-cli/test/build.ts (hunk 16 above)
Comment (nisheetjain) [inline src/taco-cli/test/build.ts]: done

File: src/taco-cli/test/build.ts (hunk 17 above)
Comment (nisheetjain) [inline src/taco-cli/test/build.ts]: done

File: src/taco-cli/test/build.ts (hunk 18 above)
Comment (nisheetjain) [inline src/taco-cli/test/build.ts]: makes sense. done

File: src/remotebuild/lib/darwin/darwinCerts.ts (hunk 1 above)
Comment (nisheetjain) [inline src/remotebuild/lib/darwin/darwinCerts.ts]: done

File: src/remotebuild/lib/darwin/darwinCerts.ts (hunk 2 above)
Comment (nisheetjain) [inline src/remotebuild/lib/darwin/darwinCerts.ts]: done

File: src/remotebuild/lib/server.ts (hunk 3 above)
Comment (nisheetjain) [inline src/remotebuild/lib/server.ts]: done

File: src/remotebuild/test/darwinCerts.ts (hunk 4 above)
Comment (nisheetjain) [inline src/remotebuild/test/darwinCerts.ts]: wouldn't prefer touching this part in a separate checkin if needed

File: src/remotebuild/test/darwinCerts.ts (hunk 5 above)
Comment (nisheetjain) [inline src/remotebuild/test/darwinCerts.ts]: done

File: src/remotebuild/test/darwinCerts.ts (hunk 6 above)
Comment (nisheetjain) [inline src/remotebuild/test/darwinCerts.ts]: done

File: src/remotebuild/test/server.ts (hunk 7 above)
Comment (nisheetjain) [inline src/remotebuild/test/server.ts]: same as before..

File: src/taco-cli/cli/build.ts (hunk 8 above)
Comment (nisheetjain) [inline src/taco-cli/cli/build.ts]: done

File: src/taco-cli/cli/utils/cordovaHelper.ts (hunk 11 above)
Comment (nisheetjain) [inline src/taco-cli/cli/utils/cordovaHelper.ts]: these ones are tricky as they are actually defined as types here..

File: src/taco-cli/cli/utils/cordovaHelper.ts (hunk 12 above)
Comment (nisheetjain) [inline src/taco-cli/cli/utils/cordovaHelper.ts]: same as above

File: src/taco-cli/cli/create.ts [hunk 19]
```diff
@@ -78,23 +78,23 @@ class Create extends commands.TacoCommandBase {
...
+            .then(function(templateUsed: string): Q.Promise<any> {
                 templateDisplayName = templateUsed;
 
-                var kitProject = self.isKitProject();
+                var kitProject: boolean = self.isKitProject();
                 var valueToSerialize: string = kitProject ? self.commandParameters.data.options["kit"] : self.commandParameters.data.options["cordova"];
 
                 return projectHelper.createTacoJsonFile(self.commandParameters.cordovaParameters.projectPath, kitProject, valueToSerialize);
             })
-            .then(function (): Q.Promise<any> {
```
Comment (meenakb) [inline src/taco-cli/cli/create.ts]: Looks like you the space between "function" and "()" has been deleted in some files but not in many other files. Lets be consistent here.

File: src/remotebuild/test/server.ts [hunk 20]
```diff
@@ -122,7 +122,7 @@ describe("server", function (): void {
             fs.existsSync(path.join(certsDir, "ca-key.pem")).should.be.ok;
             fs.existsSync(path.join(certsDir, "ca-cert.pem")).should.be.ok;
         }).then(function (): Q.Promise<{}> {
-            var deferred = Q.defer();
+            var deferred: Q.Deferred<any> = Q.defer();
             // Disabling strict SSL for unit testing
             request.get({ url: "https://localhost:3000", strictSSL: false }, function (error: any, response: any, body: any): void {
```
Comment (digeff) [inline src/remotebuild/test/server.ts]: According to this file: https://github.com/borisyankov/DefinitelyTyped/blob/master/request/request.d.ts it seems that response is a http.IncomingMessage

File: src/taco-cli/cli/remote.ts [hunk 21]
```diff
@@ -208,11 +208,11 @@ class Remote extends commands.TacoCommandBase {
...
     private static queryUserForRemoteConfig(): Q.Promise<{ host: string; port: number; pin: number }> {
-        var hostPromise = Q.defer<{ host: string }>();
-        var portPromise = Q.defer<{ host: string; port: number }>();
-        var pinPromise = Q.defer<{ host: string; port: number; pin: number }>();
+        var hostPromise: Q.Deferred<{ host: string }> = Q.defer<{ host: string }>();
+        var portPromise: Q.Deferred<{ host: string; port: number }> = Q.defer<{ host: string; port: number }>();
+        var pinPromise: Q.Deferred<{ host: string; port: number; pin: number }> = Q.defer<{ host: string; port: number; pin: number }>();
 
-        var cliSession = Remote.cliSession ? Remote.cliSession : readline.createInterface({ input: process.stdin, output: process.stdout });
+        var cliSession: IcliSession = Remote.cliSession ? Remote.cliSession : readline.createInterface({ input: process.stdin, output: process.stdout });
```
Comment (meenakb) [inline src/taco-cli/cli/remote.ts]: Not part of this change but per our coding guidelines, the name of this interface should be "ICliSession" not "IcliSession"

File: src/taco-cli/test/utils/httpServerFunction.ts [hunk 22]
```diff
@@ -0,0 +1,15 @@
...
+ *******************************************************
+ */
+
+/// <reference path="../../../typings/node.d.ts" />
+
+interface IHttpServerFunction {
+    (request: NodeJSHttp.ServerRequest, response: NodeJSHttp.ServerResponse): void;
+};
+
+export = IHttpServerFunction;
```
Comment (meenakb) [inline src/taco-cli/test/utils/httpServerFunction.ts]: Seems like a good candidate for a ".d.ts" file under our "typings" folder.

**Comment by meenakb** [comment]
:+1:  with comments above.

File: src/taco-cli/cli/create.ts (hunk 19 above)
Comment (nisheetjain) [inline src/taco-cli/cli/create.ts]: done

File: src/taco-cli/cli/remote.ts (hunk 21 above)
Comment (nisheetjain) [inline src/taco-cli/cli/remote.ts]: good catch, done

File: src/taco-cli/test/utils/httpServerFunction.ts (hunk 22 above)
Comment (nisheetjain) [inline src/taco-cli/test/utils/httpServerFunction.ts]: left it as is, since this is test specific. sort of internal for the tests. if we move it to typedef, it suddenly becomes public to whole code..
